src/precomputed_quantities.jl

```diff
@@ -3,49 +3,124 @@
+            Geometry.Contravariant3Axis(),
+            ᶠwinterp(ᶜYc.ρ[colidx] * ᶜJ[colidx], CT123(ᶜYc.uₕ[colidx])) +
+            CT123(ᶠw[colidx]),
+        )
+        compute_kinetic!(ᶜK[colidx], ᶜYc.uₕ[colidx], ᶠw[colidx])
+    end
+end
+
+"""
+    thermo_state(thermo_params; ρ, p, θ, e_int, q_tot, q_pt)
+
+Constructs a thermodynamic state based on the supplied keyword arguments.
+"""
+function thermo_state(
+    thermo_params;
+    ρ = nothing,
+    p = nothing,
+    θ = nothing,
+    e_int = nothing,
+    q_tot = nothing,
+    q_pt = nothing,
+)
+    get_ts(ρ::Real, ::Nothing, θ::Real, ::Nothing, ::Nothing, ::Nothing) =
+        TD.PhaseDry_ρθ(thermo_params, ρ, θ)
+    get_ts(ρ::Real, ::Nothing, θ::Real, ::Nothing, q_tot::Real, ::Nothing) =
+        TD.PhaseEquil_ρθq(thermo_params, ρ, θ, q_tot)
+    get_ts(ρ::Real, ::Nothing, θ::Real, ::Nothing, ::Nothing, q_pt) =
+        TD.PhaseNonEquil_ρθq(thermo_params, ρ, θ, q_pt)
+    get_ts(ρ::Real, ::Nothing, ::Nothing, e_int::Real, ::Nothing, ::Nothing) =
+        TD.PhaseDry_ρe(thermo_params, ρ, e_int)
+    get_ts(ρ::Real, ::Nothing, ::Nothing, e_int::Real, q_tot::Real, ::Nothing) =
+        TD.PhaseEquil_ρeq(thermo_params, ρ, e_int, q_tot)
+    get_ts(ρ::Real, ::Nothing, ::Nothing, e_int::Real, ::Nothing, q_pt) =
+        TD.PhaseNonEquil_ρeq(thermo_params, ρ, e_int, q_pt)
+    get_ts(::Nothing, p::Real, θ::Real, ::Nothing, ::Nothing, ::Nothing) =
+        TD.PhaseDry_pθ(thermo_params, p, θ)
+    get_ts(::Nothing, p::Real, θ::Real, ::Nothing, q_tot::Real, ::Nothing) =
+        TD.PhaseEquil_pθq(thermo_params, p, θ, q_tot)
+    get_ts(::Nothing, p::Real, θ::Real, ::Nothing, ::Nothing, q_pt) =
+        TD.PhaseNonEquil_pθq(thermo_params, p, θ, q_pt)
+    get_ts(::Nothing, p::Real, ::Nothing, e_int::Real, ::Nothing, ::Nothing) =
+        TD.PhaseDry_pe(thermo_params, p, e_int)
+    get_ts(::Nothing, p::Real, ::Nothing, e_int::Real, q_tot::Real, ::Nothing) =
+        TD.PhaseEquil_peq(thermo_params, p, e_int, q_tot)
+    get_ts(::Nothing, p::Real, ::Nothing, e_int::Real, ::Nothing, q_pt) =
+        TD.PhaseNonEquil_peq(thermo_params, p, e_int, q_pt)
+    return get_ts(ρ, p, θ, e_int, q_tot, q_pt)
+end
+
+function ts(Yc, K, Φ, thermo_params, energy_form, moisture_model)
+    energy_var = if energy_form isa PotentialTemperature
+        (; θ = Yc.ρθ / Yc.ρ)
+    elseif energy_form isa TotalEnergy
+        e_tot = Yc.ρe_tot / Yc.ρ
+        (; e_int = e_tot - K - Φ)
+    end
+    moisture_var = if moisture_model isa DryModel
+        (;)
+    else
+        q_tot = Yc.ρq_tot / Yc.ρ
+        if moisture_model isa EquilMoistModel
+            (; q_tot)
+        elseif moisture_model isa NonEquilMoistModel
+            q_liq = Yc.ρq_liq / Yc.ρ
+            q_ice = Yc.ρq_ice / Yc.ρ
+            (; q_pt = TD.PhasePartition(q_tot, q_liq, q_ice))
+        end
+    end
+    return thermo_state(thermo_params; ρ = Yc.ρ, energy_var..., moisture_var...)
+end
+
+function set_precomputed_quantities!(Y, p, t)
```

Comment:
Was wondering if we could go for a more descriptive name here? For example `set_thermo_state_and_pressure`. Though I guess we rick it becoming outdated if we add more things there